kube-core/src/conversion/types.rs

```diff
@@ -0,0 +1,212 @@
+}
+
+/// Part of ConversionReview which is set on input (i.e. generated by apiserver)
+#[derive(Serialize, Deserialize)]
+pub struct ConversionRequest {
+    /// [`TypeMeta`] of the [`ConversionReview`] this response was created from
+    ///  
+    /// This field dopied from the corresponding [`ConversionReview`].
+    /// It is not part of the Kubernetes API, it's consumed only by `kube`.
+    #[serde(skip)]
+    pub types: Option<TypeMeta>,
+    /// Random uid uniquely identifying this conversion call
+    pub uid: String,
+    /// The API group and version the objects should be converted to
+    #[serde(rename = "desiredAPIVersion")]
+    pub desired_api_version: String,
+    /// The list of objects to convert
+    ///
+    /// Note that list may contain one or more objects, in one or more versions.
+    // This field uses raw Value instead of Object/DynamicObject to simplify
+    // further downcasting.
+    pub objects: Vec<serde_json::Value>,
+}
+
+impl ConversionRequest {
+    /// Extracts request from the [`ConversionReview`]
+    pub fn from_review(review: ConversionReview) -> Result<Self, ConvertConversionReviewError> {
+        ConversionRequest::try_from(review)
+    }
+}
+
+impl TryFrom<ConversionReview> for ConversionRequest {
+    type Error = ConvertConversionReviewError;
+
+    fn try_from(review: ConversionReview) -> Result<Self, Self::Error> {
+        match review.request {
+            Some(mut req) => {
+                req.types = Some(review.types);
+                Ok(req)
+            }
+            None => Err(ConvertConversionReviewError),
+        }
+    }
+}
+
+/// Part of ConversionReview which is set on output (i.e. generated by conversion webhook)
+#[derive(Serialize, Deserialize)]
+pub struct ConversionResponse {
+    /// [`TypeMeta`] of the [`ConversionReview`] this response was derived from
+    ///  
+    /// This field is copied from the corresponding [`ConversionRequest`].
+    /// It is not part of the Kubernetes API, it's consumed only by `kube`.
+    #[serde(skip)]
+    pub types: Option<TypeMeta>,
+    /// Copy of .request.uid
+    pub uid: String,
+    /// Outcome of the conversion operation
+    ///
+    /// Success: all objects were successfully converted
+    /// Failure: at least one object could not be converted.
+    /// It is recommended that conversion fails as rare as possible.
+    pub result: Status,
+    /// Converted objects
+    ///
+    /// This field should contain objects in the same order as in the request
+    /// Should be empty if conversion failed.
+    #[serde(rename = "convertedObjects")]
+    #[serde(deserialize_with = "parse_converted_objects")]
+    pub converted_objects: Vec<serde_json::Value>,
+}
+
+fn parse_converted_objects<'de, D>(de: D) -> Result<Vec<serde_json::Value>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum Helper {
+        List(Vec<serde_json::Value>),
+        Null(()),
+    }
```

Comment:
so this is basically accounting for an omitted `converted_objects` or `converted_objects: null` and the normal `converted_objects:[...]`? could this not be handled with a `#[serde(default)]` instead?

Reply:
```
thread 'conversion::types::tests::simple_request_parses' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: null, expected a sequence", line: 50, column: 32)', kube-core/src/conversion/types.rs:207:49
```
AFAIU, `#[serde(default)]` applies when value is completely missing, but apiserver (at least the one I've tested my changes on) sets `null`.

Reply:
Well, then we are back to the same question of whether or not `converted_objects` can be `null`.
**If** your `simple.json` is actually retrieved from the apiserver, then I would rather have `Option<Vec<..>>` because then we can use the simpler `#[serde(default)]`.
...but, neither the [source](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types.go#L474-L478), nor the [protobufs](https://github.com/kube-rs/k8s-pb/blob/d60271637aac5bb854d529f99d5e6a23154acd06/k8s-pb/src/apiextensions_apiserver/pkg/apis/apiextensions/v1/mod.rs#L29) list it as optional, so i am curious whether your test case is actually real. if it is not, i would not bother trying to handle it, if it is, then it's probably an upstream Kubernetes bug.

Reply:
> If your simple.json is actually retrieved from the apiserver

Yes. I've written a web server which logs incoming requests (without deserializing to Rust structs), and specified it as a conversion webhook.

I *guess* [here](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apiserver/conversion/webhook_converter.go#L156) a `nil` slice is assigned, which later gets converted into `null`.

I guess I should try the same on a vanilla `kind` cluster and see what happens (because I used a managed k8s service, hence it is *possible* its admission code was patched).

Reply:
Well, the source hasn't changed in 3 years so if you are seeing it, we will likely see it from others. So, it probably makes sensi to keep this more defensive deserializer in here. It's not a big deal, and it even makes the user expectation match the go structs.

Reply:
I confirmed this behavior using kind v0.15.0 and vanilla k8s v1.25.0.
I'm marking this conversation as resolved: while hacks on our side are unfortunate, we can't avoid them right now.